text/0000-custom-dst.md

```diff
@@ -0,0 +1,479 @@
+pointers to `DynamicallySized` types:
+
+```rust
+mod core::raw {
+    pub fn from_raw_parts<T: ?Sized>(
+        ptr: *const (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *const T;
+
+    pub fn from_raw_parts_mut<T: ?Sized>(
+        ptr: *mut (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *mut T;
+
+    pub fn metadata<T: ?Sized>(
+        ptr: *const T,
+    ) -> <T as DynamicallySized>::Metadata;
+}
+```
+
+and we will introduce two functions into `core::mem`,
+to help people write types with Flexible Array Members:
+
+```rust
+mod core::mem {
+    pub fn size_of_header<T: ?DynamicallySized>() -> usize;
+    pub fn align_of_header<T: ?DynamicallySized>() -> usize;
+}
+```
+
+These functions return the size and alignment of the header of a type;
+or, the minimum possible size and alignment, in other words.
+For existing `Sized` types, they are equivalent to `size_of` and `align_of`,
+and for existing DSTs,
+
+```rust
+assert_eq!(size_of_header::<[T]>(), 0);
+assert_eq!(align_of_header::<[T]>(), align_of::<T>());
+assert_eq!(size_of_header::<dyn Trait>(), 0);
+assert_eq!(align_of_header::<dyn Trait>(), 1);
+
+// on 64-bit
+struct RcBox<T: ?Sized> {
+  strong: Cell<usize>,
+  weak: Cell<usize>,
+  value: T,
+}
+assert_eq!(size_of_header::<RcBox<dyn Trait>>(), 16);
+assert_eq!(align_of_header::<RcBox<dyn Trait>>(), 8);
+```
+
+Note that this is a minimum - this means that for `extern type`s,
+they return `0` and `1` respectively.
+
+Notes:
+  - names of the above functions should be bikeshed
+  - `extern type`s do not implement `DynamicallySized`, although in theory one
+    could choose to implement the trait for them
+    (that usecase is not supported by this RFC).
+  - `DynamicallySized` is a new trait in the `Sized` hierarchy
+    - this means that, by default, `T` implies `T: Sized + DynamicallySized`,
+      unless one removes that bound explicitly with `T: ?DynamicallySized`
+  - `T: ?DynamicallySized` bounds imply a `T: ?Sized` bound,
+    since `T: Sized` implies `T: DynamicallySized`
+  - `T: ?Sized` bounds do not remove the `T: DynamicallySized` requirement.
+
+We will also change `CStr` to have the implementation from above.
+
+On an ABI level, we promise that pointers to any type with
+
+```rust
+size_of::<Metadata>() == 0
+&& align_of::<Metadata>() <= align_of::<*const ()>()
+```
+
+are ABI compatible with a C pointer - this is important,
+since we want to be able to write:
+
+```rust
+extern "C" {
+    fn printf(fmt: &CStr, ...) -> c_int;
+}
+```
+
+Unfortunately, we won't be able to change existing declarations in `libc`
+without a new major version.
+
+`as` casts continue to allow
+
+```rust
+fn cast_to_thin<T: ?Sized, U: Sized>(t: *const T) -> *const U {
+    t as *const U
+}
+```
+
+so we do not introduce any new functions to access the pointer part
+of the thick pointer.
+
+The `DynamicallySized` trait may be implemented for any struct or union type
```

Comment:
I wonder if these are allowed (if not, how the compiler can disallow these):

```rust
struct S;

unsafe impl<'a> DynamicallySized for &'a S { ... }
unsafe impl DynamicallySized for Box<S> { ... } // Box is a struct
```

Reply:
I assume this would be similar to how `Drop` or `CoerceUnsized` work - the compiler can just have special-case code.

And in `Box`'s case, probably the simplest thing is to say that `DynamicallySized` can only be implemented in the same crate as the type it's implemented on.

Another option is to require a maximally generic impl, e.g. like how `Drop` can't have more bounds or more specific types than the type itself, which would mean you'd be forced to write:
```
`unsafe impl<T: ?Sized> DynamicallySized for Box<T> { ... }
```
which would then not work outside of liballoc.

Reply:
@ubsan Could you amend the RFC to mention this restriction?

Reply:
Um... `Box<T>` is always statically sized?